pkg/istio-agent/health/health_probers.go

```diff
@@ -0,0 +1,156 @@
+// limitations under the License.
+
+package health
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"os/exec"
+	"time"
+
+	"istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pkg/test/echo/common/scheme"
+	"istio.io/pkg/log"
+)
+
+var healthCheckLog = log.RegisterScope("healthcheck", "Health Checks performed by Istio-Agent", 0)
+
+type Prober interface {
+	// Probe will healthcheck and return whether or not the target is healthy.
+	// If an error returned is not nil, it is assumed that the process could
+	// not complete, and Probe() was unable to determine whether or not the
+	// target was healthy.
+	Probe(timeout time.Duration) (bool, error)
+}
+
+type HTTPProber struct {
+	Config *v1alpha3.HTTPHealthCheckConfig
+}
+
+// HttpProber_Probe will return whether or not the target is healthy (true -> healthy)
+// 	by making an HTTP Get response.
+func (h *HTTPProber) Probe(timeout time.Duration) (bool, error) {
+	client := &http.Client{
+		Timeout: timeout,
+	}
+	// modify transport if scheme is http
+	if h.Config.Scheme == string(scheme.HTTP) {
+		client.Transport = &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		}
+	}
+	// transform crd into net http header
+	headers := make(http.Header)
+	for _, val := range h.Config.HttpHeaders {
+		// net.httpHeaders value is a []string but uses only index 0
+		headers[val.Name] = []string{val.Value}
+	}
+	// GET scheme://ip:port/path
+	targetURL, err := url.Parse(fmt.Sprintf("%s://localhost:%v/%s", h.Config.Scheme, h.Config.Port, h.Config.Path))
+	if err != nil {
+		healthCheckLog.Errorf("unable to parse url: %v", err)
+		return false, err
+	}
+	req, err := http.NewRequest("GET", targetURL.String(), nil)
+	if err != nil {
+		return false, err
+	}
+	req.Header = headers
+	if headers.Get("Host") != "" {
+		req.Host = headers.Get("Host")
+	}
+	res, err := client.Do(req)
+	// if we were unable to connect, count as failure
+	if err != nil {
+		healthCheckLog.Infof("Health Check failed for %v: %v", targetURL.String(), err)
+		return false, nil
+	}
+	defer func() {
+		err = res.Body.Close()
+		if err != nil {
+			healthCheckLog.Error(err)
+		}
+	}()
+	// from [200,400)
+	if res.StatusCode >= http.StatusOK && res.StatusCode < http.StatusBadRequest {
+		healthCheckLog.Infof("Health check succeeded for %v", targetURL.String())
+		return true, nil
+	}
+	return false, nil
+}
+
+type TCPProber struct {
+	Config *v1alpha3.TCPHealthCheckConfig
+}
+
+func (t *TCPProber) Probe(timeout time.Duration) (bool, error) {
+	// if we cant connect, count as fail
+	conn, err := net.DialTimeout("tcp", fmt.Sprintf("%s:%v", t.Config.Host, t.Config.Port), timeout)
+	if err != nil {
+		return false, nil
+	}
+	err = conn.Close()
+	if err != nil {
+		healthCheckLog.Errorf("Unable to close TCP Socket: %v", err)
+	}
+	return true, nil
+}
+
+type ExecProber struct {
+	Config *v1alpha3.ExecHealthCheckConfig
+}
+
+func (e *ExecProber) Probe(timeout time.Duration) (bool, error) {
```

Comment:
Does this have the potential to trigger a panic in our code? I am not sure how exec.Cmd handles it. But we must not panic here even if the user command does crash.

Reply:
i dont know. will have to look into it